ext/telemetry/lib.rs

```diff
@@ -974,188 +1057,545 @@ fn owned_string<'s>(
+  span: v8::Local<'_, v8::Value>,
   key: v8::Local<'s, v8::Value>,
   value: v8::Local<'s, v8::Value>,
 ) {
-  if let Some(temporary_span) = state.try_borrow_mut::<TemporarySpan>() {
-    temporary_span.0.attributes.reserve_exact(
-      (capacity as usize)
-        .saturating_sub(temporary_span.0.attributes.capacity()),
-    );
-    attr!(scope, temporary_span.0.attributes => temporary_span.0.dropped_attributes_count, key, value);
+  let Some(span) =
+    deno_core::_ops::try_unwrap_cppgc_object::<OtelSpan>(scope, span)
+  else {
+    return;
+  };
+  let mut state = span.0.borrow_mut();
+  if let OtelSpanState::Recording(span) = &mut **state {
+    attr!(scope, span.attributes => span.dropped_attributes_count, key, value);
   }
 }
 
 #[op2(fast)]
 fn op_otel_span_attribute2<'s>(
   scope: &mut v8::HandleScope<'s>,
-  state: &mut OpState,
-  #[smi] capacity: u32,
+  span: v8::Local<'_, v8::Value>,
   key1: v8::Local<'s, v8::Value>,
   value1: v8::Local<'s, v8::Value>,
   key2: v8::Local<'s, v8::Value>,
   value2: v8::Local<'s, v8::Value>,
 ) {
-  if let Some(temporary_span) = state.try_borrow_mut::<TemporarySpan>() {
-    temporary_span.0.attributes.reserve_exact(
-      (capacity as usize)
-        .saturating_sub(temporary_span.0.attributes.capacity()),
-    );
-    attr!(scope, temporary_span.0.attributes => temporary_span.0.dropped_attributes_count, key1, value1);
-    attr!(scope, temporary_span.0.attributes => temporary_span.0.dropped_attributes_count, key2, value2);
+  let Some(span) =
+    deno_core::_ops::try_unwrap_cppgc_object::<OtelSpan>(scope, span)
+  else {
+    return;
+  };
+  let mut state = span.0.borrow_mut();
+  if let OtelSpanState::Recording(span) = &mut **state {
+    attr!(scope, span.attributes => span.dropped_attributes_count, key1, value1);
+    attr!(scope, span.attributes => span.dropped_attributes_count, key2, value2);
   }
 }
 
 #[allow(clippy::too_many_arguments)]
 #[op2(fast)]
 fn op_otel_span_attribute3<'s>(
   scope: &mut v8::HandleScope<'s>,
-  state: &mut OpState,
-  #[smi] capacity: u32,
+  span: v8::Local<'_, v8::Value>,
   key1: v8::Local<'s, v8::Value>,
   value1: v8::Local<'s, v8::Value>,
   key2: v8::Local<'s, v8::Value>,
   value2: v8::Local<'s, v8::Value>,
   key3: v8::Local<'s, v8::Value>,
   value3: v8::Local<'s, v8::Value>,
 ) {
-  if let Some(temporary_span) = state.try_borrow_mut::<TemporarySpan>() {
-    temporary_span.0.attributes.reserve_exact(
-      (capacity as usize)
-        .saturating_sub(temporary_span.0.attributes.capacity()),
-    );
-    attr!(scope, temporary_span.0.attributes => temporary_span.0.dropped_attributes_count, key1, value1);
-    attr!(scope, temporary_span.0.attributes => temporary_span.0.dropped_attributes_count, key2, value2);
-    attr!(scope, temporary_span.0.attributes => temporary_span.0.dropped_attributes_count, key3, value3);
+  let Some(span) =
+    deno_core::_ops::try_unwrap_cppgc_object::<OtelSpan>(scope, span)
+  else {
+    return;
+  };
```

Comment:
Maybe we should support something like `Option<OtelSpan>` in the macro

Reply:
We actually do support that - the problem is that right now an op can take on of cppgc object, or `HandleScope`, not both - I have a fix in a PR somewhere in core - I think the one that changes string handling.